File: gimbal/src/main/java/com/commit451/gimbal/Gimbal.java
package com.commit451.gimbal;

import android.app.Activity;
import android.content.pm.ActivityInfo;
import android.graphics.Point;
import android.hardware.SensorEvent;
import android.view.Display;
import android.view.Surface;

import java.lang.ref.WeakReference;

/**
 * Tool to help out with rotation and orientation based events
 */
public class Gimbal {

    private WeakReference<Activity> mActivityWeakReference;
    private Point mNormalizationPoint;
    private Point mDisplayPoint;

    /**
     * Create a new instance of Gimbal, using the context and values of the activity
     * @param activity the activity to get values from and potentially modify
     */
    public Gimbal(Activity activity) {
        mActivityWeakReference = new WeakReference<>(activity);
    }

    /**
     * Locks the activity to the orientation the user was in when they entered the activity.
     * @return true if lock was successful, false if it was not
     */
    public boolean lock() {
        Activity activity = mActivityWeakReference.get();
        if (activity == null) {
            return false;
        }
        Display display = activity.getWindowManager().getDefaultDisplay();
        int rotation = display.getRotation();
        int height;
        int width;
        if (mDisplayPoint == null) {
            mDisplayPoint = new Point();
        }
        display.getSize(mDisplayPoint);
        height = mDisplayPoint.y;
        width = mDisplayPoint.x;
        switch (rotation) {
            case Surface.ROTATION_90:
                if (width > height) {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
                }
                else {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT);
                }
                break;
            case Surface.ROTATION_180:
                if (height > width) {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT);
                }
                else {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE);
                }
                break;
            case Surface.ROTATION_270:
                if (width > height) {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE);
                }
                else {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                }
                break;
            case Surface.ROTATION_0:
            default :
                if (height > width) {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                }
                else {
                    activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE);
                }
        }
        return true;
    }

    /**
     * Unlock the previously locked activity
     * @return true if successfully unlocked. False otherwise
     */
    public boolean unlock() {
        Activity activity = mActivityWeakReference.get();
        if (activity == null) {
            return false;
        }
        activity.setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED);
        return true;
    }

    /**
     * Normalize a {@link SensorEvent} from the gravity sensor for the current orientation.
     * @param event the {@link SensorEvent} which will be normalized, typically acquired in a {@link android.hardware.SensorEventListener} event
     * @return true if event has been normalized. False otherwise
     */
    public boolean normalizeGravityEvent(SensorEvent event) {
        Activity activity = mActivityWeakReference.get();
        if (activity == null) {
            return false;
        }
        Display display = activity.getWindow().getWindowManager().getDefaultDisplay();
        int rotation = display.getRotation();
        int height;
        int width;
        if (mNormalizationPoint == null) {
            mNormalizationPoint = new Point();
        }
        display.getSize(mNormalizationPoint);
        height = mNormalizationPoint.y;
        width = mNormalizationPoint.x;
        float x;
        float y;
        switch (rotation) {
            case Surface.ROTATION_90:
                if (width > height) {
                    //landscape
                    x = -event.values[1];
                    y = event.values[0];
                }
                else {
                    //reverse portrait
                    x = -event.values[0];
                    y = -event.values[1];
                }
                break;
            case Surface.ROTATION_180:
                if (height > width) {
                    //reverse portrait
                    x = -event.values[0];
                    y = -event.values[1];
                }
                else {
                    //reverse landscape
                    x = event.values[1];
                    y = -event.values[0];
                }
                break;
            case Surface.ROTATION_270:
                if (width > height) {
                    //reverse landscape
                    x = event.values[1];
                    y = -event.values[0];
                }
                else {
                    //portrait
                    x = event.values[0];
                    y = event.values[1];
                }
                break;
            case Surface.ROTATION_0:
            default :
                if (height > width) {
                    //portrait
                    x = event.values[0];
                    y = event.values[1];
                }
                else {
                    //landscape
                    x = -event.values[1];
                    y = event.values[0];
                }
        }
        event.values[0] = x;
        event.values[1] = y;
        return true;
    }
}

File: gimbal/src/main/java/com/commit451/gimbal/Gimbal.kt
package com.commit451.gimbal

import android.app.Activity
import android.content.pm.ActivityInfo
import android.graphics.Point
import android.hardware.SensorEvent
import android.view.Surface

import java.lang.ref.WeakReference

/**
 * Tool to help out with rotation and orientation based events
 */
class Gimbal
/**
 * Create a new instance of Gimbal, using the context and values of the activity
 * @param activity the activity to get values from and potentially modify
 */
(activity: Activity) {

    private val activityWeakReference: WeakReference<Activity> = WeakReference(activity)
    private val normalizationPoint: Point by lazy(LazyThreadSafetyMode.NONE) {
        Point()
    }
    private val displayPoint: Point by lazy(LazyThreadSafetyMode.NONE) {
        Point()
    }

    /**
     * Locks the activity to the orientation the user was in when they entered the activity.
     * @return true if lock was successful, false if it was not
     */
    fun lock(): Boolean {
        val activity = activityWeakReference.get() ?: return false
        val display = activity.windowManager.defaultDisplay
        val rotation = display.rotation
        display.getSize(displayPoint)
        val height = displayPoint.y
        val width = displayPoint.x
        when (rotation) {
            Surface.ROTATION_90 -> if (width > height) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            } else {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT
            }
            Surface.ROTATION_180 -> if (height > width) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_PORTRAIT
            } else {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE
            }
            Surface.ROTATION_270 -> if (width > height) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_REVERSE_LANDSCAPE
            } else {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            }
            Surface.ROTATION_0 -> if (height > width) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            } else {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            }
            else -> if (height > width) {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
            } else {
                activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
            }
        }
        return true
    }

    /**
     * Unlock the previously locked activity
     * @return true if successfully unlocked. False otherwise
     */
    fun unlock(): Boolean {
        val activity = activityWeakReference.get() ?: return false
        activity.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
        return true
    }

    /**
     * Normalize a [SensorEvent] from the gravity sensor for the current orientation.
     * @param event the [SensorEvent] which will be normalized, typically acquired in a [android.hardware.SensorEventListener] event
     * @return true if event has been normalized. False otherwise
     */
    fun normalizeGravityEvent(event: SensorEvent): Boolean {
        val activity = activityWeakReference.get() ?: return false
        val display = activity.window.windowManager.defaultDisplay
        val rotation = display.rotation
        display.getSize(normalizationPoint)
        val height = normalizationPoint.y
        val width = normalizationPoint.x
        val x: Float
        val y: Float
        when (rotation) {
            Surface.ROTATION_90 -> if (width > height) {
                //landscape
                x = -event.values[1]
                y = event.values[0]
            } else {
                //reverse portrait
                x = -event.values[0]
                y = -event.values[1]
            }
            Surface.ROTATION_180 -> if (height > width) {
                //reverse portrait
                x = -event.values[0]
                y = -event.values[1]
            } else {
                //reverse landscape
                x = event.values[1]
                y = -event.values[0]
            }
            Surface.ROTATION_270 -> if (width > height) {
                //reverse landscape
                x = event.values[1]
                y = -event.values[0]
            } else {
                //portrait
                x = event.values[0]
                y = event.values[1]
            }
            Surface.ROTATION_0 -> if (height > width) {
                //portrait
                x = event.values[0]
                y = event.values[1]
            } else {
                //landscape
                x = -event.values[1]
                y = event.values[0]
            }
            else -> if (height > width) {
                x = event.values[0]
                y = event.values[1]
            } else {
                x = -event.values[1]
                y = event.values[0]
            }
        }
        event.values[0] = x
        event.values[1] = y
        return true
    }
}
